### PR TITLE
Patch 2478

### LINKED ISSUE #2478: TransformerEncoder is not causal
### Add Link

https://pytorch.org/tutorials/beginner/transformer_tutorial.html

image

for language modeling， src_mask should be mask future words

### Describe the bug

is there anything wrong?

### Describe your environment

 colab

cc @pytorch/team-text-core @Nayef211 @sekyondaMeta @svekars @carljparker @NicolasHug @kit1980 @subramen

### PR DESCRIPTION
**Pull request by ahoblitz**
Fixes #2478

## Description
Use `nn.Transformer.generate_square_subsequent_mask` when `src_mask` is `None` (Had to run to around ~30 epochs to get somewhat sensible output
Screenshot 2023-11-07 at 7 40 55 PM

## Checklist

- [X] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [X] Only one issue is addressed in this pull request
- [X] Labels from the issue that this PR is fixing are added to this pull request
- [X] No unnecessary issues are included into this pull request.

cc @pytorch/team-text-core @Nayef211 @sekyondaMeta @svekars @carljparker @NicolasHug @kit1980 @subramen

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
@drisspg can you double check this please?

**Comment by drisspg** [comment]
I think I am okay with this but can you add some to the description saying that this is doing Causal Language Modeling and maybe a comment as to why we need this mask

**Comment by ahoblitz** [comment]
@albanD @drisspg I added some text description to explain the need for generating a square causal mask for the sequence

File: beginner_source/transformer_tutorial.py [hunk 1]
```diff
@@ -41,8 +41,10 @@
 # Along with the input sequence, a square attention mask is required because the
 # self-attention layers in ``nn.TransformerDecoder`` are only allowed to attend
 # the earlier positions in the sequence. For the language modeling task, any
-# tokens on the future positions should be masked. To produce a probability
-# distribution over output words, the output of the ``nn.TransformerEncoder``
+# tokens on the future positions should be masked.  This masking, combined with fact that 
+# the output embeddings are offset with later positions ensures that the
+# predictions for position i can depend only on the known outputs at positions less than i.
+# To produce a probability  distribution over output words, the output of the ``nn.TransformerEncoder``
```
Comment (drisspg) [inline beginner_source/transformer_tutorial.py:47]: Nit: can you update one more spot and at the very top of this can you change
```
# In this tutorial, we train a ``nn.TransformerEncoder`` model on a
# language modeling task. Please note that this tutorial does not cover
```
to
```
# In this tutorial, we train a ``nn.TransformerEncoder`` model on a
# causal language modeling task. Please note that this tutorial does not cover
```

Comment (ahoblitz) [reply beginner_source/transformer_tutorial.py:47]: Done

**Review by drisspg — APPROVED** [review]
